taskcat/_cli_modules/test.py

```diff
@@ -133,86 +261,31 @@ def run(  # noqa: C901
+        :param skip_upload: dont upload resources to s3
         """
-        project_root_path: Path = Path(project_root).expanduser().resolve()
-        input_file_path: Path = project_root_path / input_file
-        # pylint: disable=too-many-arguments
-        args = _build_args(enable_sig_v2, regions, GLOBAL_ARGS.profile)
-        config = Config.create(
-            project_root=project_root_path,
-            project_config_path=input_file_path,
-            args=args
-            # TODO: detect if input file is taskcat config or CloudFormation template
-        )
-        _trim_regions(regions, config)
-        _trim_tests(test_names, config)
-        boto3_cache = Boto3Cache()
-        templates = config.get_templates()
-        if skip_upload and not config.config.project.s3_bucket:
-            raise TaskCatException(
-                "cannot skip_buckets without specifying s3_bucket in config"
-            )
-        buckets = config.get_buckets(boto3_cache)
-        if not skip_upload:
-            # 1. lint
-            if not lint_disable:
-                lint = TaskCatLint(config, templates)
-                errors = lint.lints[1]
-                lint.output_results()
-                if errors or not lint.passed:
-                    raise TaskCatException("Lint failed with errors")
-            # 2. build lambdas
-            if config.config.project.package_lambda:
-                LambdaBuild(config, project_root_path)
-            # 3. s3 sync
-            stage_in_s3(buckets, config.config.project.name, config.project_root)
-        # 4. launch stacks
-        regions = config.get_regions(boto3_cache)
-        parameters = config.get_rendered_parameters(buckets, regions, templates)
-        tests = config.get_tests(templates, regions, buckets, parameters)
-        test_definition = Stacker(
-            config.config.project.name,
-            tests,
-            shorten_stack_name=config.config.project.shorten_stack_name,
+
+        test = cls(
+            test_names, regions, input_file, project_root, enable_sig_v2, minimal_output
         )
-        test_definition.create_stacks()
-        terminal_printer = TerminalPrinter(minimalist=minimal_output)
-        # 5. wait for completion
-        terminal_printer.report_test_progress(stacker=test_definition)
-        status = test_definition.status()
-        # 6. create report
+
+        test.create_stacks(skip_upload, lint_disable)
+
         report_path = Path(output_directory).resolve()
         report_path.mkdir(exist_ok=True)
         cfn_logs = _CfnLogTools()
-        cfn_logs.createcfnlogs(test_definition, report_path)
-        ReportBuilder(test_definition, report_path / "index.html").generate_report()
-        # 7. delete stacks
+        cfn_logs.createcfnlogs(test.test_definition, report_path)
+        ReportBuilder(
+            test.test_definition, report_path / "index.html"
+        ).generate_report()
```

Comment:
Only a few lines of CLI specific stuff left. Lots of overlap between running the test in python and from the CLI which is a good thing.